### PR TITLE
feat: added method fetchNumberPages that calc number of pages more faster and correctly

### PR DESCRIPTION
**Pull request by truechernyshov**
## Problem with large file. #138

Imagick too long calculating number of pages and It turned out not always correctly(see test A17_FlightPlan.pdf). Server was down with 503 error on shared hosting because on that max execuution time 30s. I suggest parsing document with Smalot\PdfParser\Parser that work quickly and correctly.

### pdf document 40 pages
old
Time: 37.35 seconds, Memory: 4.00 MB
OK (1 test, 1 assertion)
Process finished with exit code 0
new
Time: 695 ms, Memory: 33.55 MB
OK (1 test, 1 assertion)
Process finished with exit code 0

### pdf document 618 pages
old
Failed asserting that false is true.
#wrong calc number of pages 370 instead of 618
Time: 34.77 seconds, Memory: 6.00 MB
FAILURES!
Tests: 1, Assertions: 1, Failures: 1.

new
Time: 20.79 seconds, Memory: 79.75 MB
OK (1 test, 1 assertion)
Process finished with exit code 0

### CODE REVIEW

File: src/Pdf.php [hunk 1]
```diff
@@ -35,13 +36,22 @@ public function __construct(string $pdfFile)
...
 
-        $this->numberOfPages = $this->imagick->getNumberImages();
+        $this->numberOfPages = $this->fetchNumberPages($pdfFile);
 
         $this->pdfFile = $pdfFile;
     }
 
+    public function fetchNumberPages(string $pdfFile)
+    {
+        $parser = new Parser();
```
Comment (freekmurze) [inline src/Pdf.php]: What code is being executed in that package? My it's simple enough to transfer to this package, so we don't have take on a dependency.

Comment (truechernyshov) [reply src/Pdf.php]: There big class that parse pdf document. Alternatively can use unix tool pdfinfo and then run exec('phpinfo '.$filename);

**Comment by rdevroede** [comment]
Using pingImage should improve speed, as that only fetches basic info. https://www.php.net/manual/en/imagick.pingimage.php

Something like:
```
$im = new Imagick();
$im->pingImage($pdfFile);
echo $im->getNumberImages();
```

There is no simple pure PHP solution AFAIK.

Other options are to include a dependency like [TCPDI](https://github.com/pauln/tcpdi) or use an external command like pdfinfo.

**Comment by rubenvanassche** [comment]
Okay I took a look at this, I've tested the whole test suite and these we're the times it took to run the fetch pages function for each test on average:

- PDF Parser: 1.33 seconds
- Current Imagick implementation: 2.70 sconds
- Imagick implementation suggested by @rdevroede:  1.94 seconds

First: @truechernyshov, can you describe me how you could run a test that takes 34.77 seconds to run? When I'm running the current implementation it takes 30.93 seconds for the whole test suite.

Second: The failing test with the bigger pdf file does not crash on my machine, a macOS computer. Haven't tested it on a server deployed by Forge yet though.

Third: The memory usages for the whole test suite are:
- PDF Parser: 79.75 MB
- Current Imagick implementation:  4.00 MB
- Imagick implementation suggested by @rdevroede:  4.00 MB
That's a lot more of memory use for achieving the same task!

Fourth: external tools like `pdfinfo` would require users to install such tools, we want this package to work out of the box and not have to ask users to install extra software when not necessary.

Fifth: I'm not that big of a fan to add new composer dependencies when it can be omitted.

So in conclusion, I think we should replace our Imagick implementation with the one suggested by @rdevroede. The memory usage stays the same, there is a speed bump of 30% and we don't need to add another package to the composer dependencies.

@freekmurze what do you think?
@truechernyshov please could you look a bit more in why the test with number of pages is failing?

**Comment by truechernyshov** [comment]
Tested on Ubuntu 18.04, php 7.2, phpstorm console
First document was 40pages.pdf, second A17_FlightPlan.pdf
Files attached.
[40pages.pdf](https://github.com/spatie/pdf-to-image/files/4103034/40pages.pdf)
[A17_FlightPlan.pdf](https://github.com/spatie/pdf-to-image/files/4103038/A17_FlightPlan.pdf)

**Comment by rubenvanassche** [comment]
Wow, this issue completely was lost in my list of issues to do, so I've recorded the times with the two files you provided: 

- PDF Parser: 1.2453881104787
- Current Imagick implementation: 4.6706940134366
- Imagick implementation suggested by @rdevroede: 3.7051098545392

The amount is correct in the tests with the three methods, but memory usage is a lot bigger with `PDF Parser`. I don't see the benefit to add another dependency into the package that isn't our own. When converting such big pdf's a queued approach seems a better solution to this.

I will add  @rdevroede suggestion tho since it is a small fix that will speed up the process a bit.
